File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/ADPSimulator.java
package com.devingotaswitch.rankings;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.AsyncTask;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.appcompat.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.andrognito.flashbar.Flashbar;
import com.devingotaswitch.rankings.extras.FilterWithSpaceAdapter;
import com.devingotaswitch.ffrv2.R;
import com.devingotaswitch.rankings.domain.Player;
import com.devingotaswitch.rankings.domain.Rankings;
import com.devingotaswitch.rankings.domain.RosterSettings;
import com.devingotaswitch.rankings.domain.ScoringSettings;
import com.devingotaswitch.utils.Constants;
import com.devingotaswitch.utils.FlashbarFactory;
import com.devingotaswitch.utils.GeneralUtils;
import com.devingotaswitch.utils.JsoupUtils;
import com.devingotaswitch.utils.ParsingUtils;

import java.util.List;

public class ADPSimulator extends AppCompatActivity {

    private Rankings rankings;

    private Player playerToSearch;

    private static final String TAG = "ADPSimulator";

    private AutoCompleteTextView searchInput;
    private EditText roundInput;
    private EditText pickInput;
    private TextView result;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_adp_simulator);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        rankings = Rankings.init();

        Toolbar toolbar = findViewById(R.id.toolbar_adp_simulator);
        toolbar.setTitle("");
        TextView main_title = findViewById(R.id.main_toolbar_title);
        main_title.setText("ADP Simulator");
        setSupportActionBar(toolbar);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);

        final Activity act = this;
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                GeneralUtils.hideKeyboard(act);
                onBackPressed();
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();

        try {
            init();

            if (getIntent().hasExtra(Constants.PLAYER_ID)) {
                setPlayerToBeChecked(getIntent().getStringExtra(Constants.PLAYER_ID));
            }
        } catch (Exception e) {
            Log.d(TAG, "Failure setting up activity, falling back to Rankings", e);
            GeneralUtils.hideKeyboard(this);
            onBackPressed();
        }
    }

    private void init() {
        searchInput = findViewById(R.id.adp_player_selection);
        searchInput.setAdapter(null);
        final FilterWithSpaceAdapter mAdapter = GeneralUtils.getPlayerSearchAdapter(rankings, this, false, false);
        searchInput.setAdapter(mAdapter);

        searchInput.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                setPlayerToBeChecked(GeneralUtils.getPlayerIdFromSearchView(view));
            }
        });

        result = findViewById(R.id.adp_output_view);
        roundInput = findViewById(R.id.adp_pick_round);
        pickInput = findViewById(R.id.adp_pick_in_round);
        Button submit = findViewById(R.id.adp_submit_button);
        final Activity act = this;
        submit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (playerToSearch == null) {
                    FlashbarFactory.generateTextOnlyFlashbar(act, "No can do", "Invalid player, use the dropdown to pick", Flashbar.Gravity.TOP)
                            .show();
                    return;
                }
                String roundStr = roundInput.getText().toString();
                String pickStr = pickInput.getText().toString();
                if (!GeneralUtils.isInteger(roundStr) || !GeneralUtils.isInteger(pickStr)) {
                    FlashbarFactory.generateTextOnlyFlashbar(act, "No can do", "Pick/round must be provided as numbers", Flashbar.Gravity.TOP)
                            .show();
                    return;
                }
                int round = Integer.parseInt(roundStr);
                int pick = Integer.parseInt(pickStr);
                if (pick > rankings.getLeagueSettings().getTeamCount()) {
                    FlashbarFactory.generateTextOnlyFlashbar(act, "No can do", "Pick can't be higher than current league team count", Flashbar.Gravity.TOP)
                            .show();
                    return;
                }
                int overallPick = ((round - 1) * rankings.getLeagueSettings().getTeamCount()) + pick;

                GeneralUtils.hideKeyboard(act);
                getADPOddsForInput(overallPick);
            }
        });
    }

    private void getADPOddsForInput(int pick) {
        ParseADPOdds oddsParser = new ParseADPOdds(this, rankings);
        oddsParser.execute(pick, playerToSearch);
    }

    private void setPlayerToBeChecked(String id) {
        playerToSearch = rankings.getPlayer(id);
        searchInput.setText(playerToSearch.getName() + ": " + playerToSearch.getPosition() + ", " + playerToSearch.getTeamName());
    }

    private void displayResult(String output) {
        result.setText(output);
        final Activity localCopy = this;
        final String playerId = playerToSearch.getUniqueId();
        result.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                Intent intent = new Intent(localCopy, PlayerInfo.class);
                intent.putExtra(Constants.PLAYER_ID, playerId);
                startActivity(intent);
                return true;
            }
        });
        clearInputs();
    }

    private void clearInputs() {
        playerToSearch = null;
        searchInput.setText("");
        pickInput.setText("");
        roundInput.setText("");
    }

    private class ParseADPOdds extends AsyncTask<Object, Void, String> {
        private final ProgressDialog pdia;
        private final Rankings rankings;

        ParseADPOdds(ADPSimulator activity, Rankings rankings) {
            pdia = new ProgressDialog(activity);
            pdia.setCancelable(false);
            this.rankings = rankings;
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            pdia.setMessage("Please wait, doing fancy math...");
            pdia.show();
        }

        @Override
        protected void onPostExecute(String result) {
            super.onPostExecute(result);
            pdia.dismiss();
            displayResult(result);
        }

        @Override
        protected String doInBackground(Object... data) {
            int pick = (Integer) data[0];
            Player player = (Player) data[1];
            String type = "standard";
            ScoringSettings scoring = rankings.getLeagueSettings().getScoringSettings();
            RosterSettings roster = rankings.getLeagueSettings().getRosterSettings();
            if (roster.getQbCount() > 1 || (roster.getFlex() != null && roster.getQbCount() > 0 &&
                    roster.getFlex().getQbrbwrteCount() > 0)) {
                type = "2qb";
            } else if (scoring.getReceptions() > 0.0) {
                type = "ppr";
            }
            String teams = "8";
            int numTeams = rankings.getLeagueSettings().getTeamCount();
            if (numTeams >= 14 && !"2qb".equals(type)) {
                teams = "14";
            } else if (numTeams >= 12) {
                teams = "12";
            } else if (numTeams >= 10) {
                teams = "10";
            }

            String url = "https://fantasyfootballcalculator.com/scenario-calculator?format="
                    + type + "&num_teams=" + teams + "&draft_pick=" + pick;
            Log.d(TAG, url);
            ParsingUtils.init();
            String first = getPlayerADPOdds(url, player, pick);
            if (scoring.getReceptions() > 0.0 && first.contains("error") && url.contains("ppr")) {
                first = getPlayerADPOdds(url.replace("ppr", "standard"), player, pick);
            }
            return first;
        }
    }

    private String getPlayerADPOdds(String url, Player player, int pick) {
        try {
            List<String> td = JsoupUtils.parseURLWithUA(url,
                    "table.table td");
            for (int i = 0; i < td.size(); i+=6) {
                String possibleName = ParsingUtils.normalizeNames(td.get(i));
                String possiblePos = td.get(i+1);
                String possibleTeam = ParsingUtils.normalizeTeams(td.get(i+2));
                if (possiblePos.equals(player.getPosition()) && possibleTeam.equals(player.getTeamName()) &&
                        (possibleName.equals(player.getName().replaceAll("\\.", "")) ||
                        possibleName.equals(player.getName()))) {
                    return "Odds " + player.getName() + " is available at pick " + pick
                            + ": " + td.get(i + 4);
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to get player adp likelihood", e);
        }
        return "An error occurred. Either the data is unavailable, or the internet may have dropped.";
    }
}

File: FFRv2/app/src/main/java/com/devingotaswitch/rankings/ADPSimulator.kt
package com.devingotaswitch.rankings

import android.app.Activity
import android.content.Intent
import android.content.pm.ActivityInfo
import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import com.andrognito.flashbar.Flashbar
import com.devingotaswitch.ffrv2.R
import com.devingotaswitch.rankings.domain.Player
import com.devingotaswitch.rankings.domain.Rankings
import com.devingotaswitch.rankings.extras.FilterWithSpaceAdapter
import com.devingotaswitch.utils.Constants
import com.devingotaswitch.utils.FlashbarFactory.generateTextOnlyFlashbar
import com.devingotaswitch.utils.GeneralUtils.getPlayerIdFromSearchView
import com.devingotaswitch.utils.GeneralUtils.getPlayerSearchAdapter
import com.devingotaswitch.utils.GeneralUtils.hideKeyboard
import com.devingotaswitch.utils.GeneralUtils.isInteger
import com.devingotaswitch.utils.JsoupUtils.parseURLWithUA
import com.devingotaswitch.utils.ParsingUtils
import com.devingotaswitch.utils.ParsingUtils.normalizeNames
import com.devingotaswitch.utils.ParsingUtils.normalizeTeams
import com.google.android.material.dialog.MaterialAlertDialogBuilder

class ADPSimulator : AppCompatActivity() {
    private var rankings: Rankings? = null
    private var playerToSearch: Player? = null
    private var searchInput: AutoCompleteTextView? = null
    private var roundInput: EditText? = null
    private var pickInput: EditText? = null
    private var result: TextView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_adp_simulator)
        requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
        rankings = Rankings.init()
        val toolbar = findViewById<Toolbar>(R.id.toolbar_adp_simulator)
        toolbar.title = ""
        val mainTitle = findViewById<TextView>(R.id.main_toolbar_title)
        mainTitle.text = "ADP Simulator"
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setDisplayShowHomeEnabled(true)
        val act: Activity = this
        toolbar.setNavigationOnClickListener { v: View? ->
            hideKeyboard(act)
            onBackPressed()
        }
    }

    public override fun onResume() {
        super.onResume()
        try {
            init()
            if (intent.hasExtra(Constants.PLAYER_ID)) {
                setPlayerToBeChecked(intent.getStringExtra(Constants.PLAYER_ID)!!)
            }
        } catch (e: Exception) {
            Log.d(TAG, "Failure setting up activity, falling back to Rankings", e)
            hideKeyboard(this)
            onBackPressed()
        }
    }

    private fun init() {
        searchInput = findViewById(R.id.adp_player_selection)
        searchInput!!.setAdapter(null)
        val mAdapter: FilterWithSpaceAdapter<*> = getPlayerSearchAdapter(rankings!!, this, hideDrafted = false, hideRankless = false)
        searchInput!!.setAdapter(mAdapter)
        searchInput!!.setOnItemClickListener { _: AdapterView<*>?, view: View?, _: Int, _: Long -> setPlayerToBeChecked(getPlayerIdFromSearchView(view!!)) }
        result = findViewById(R.id.adp_output_view)
        roundInput = findViewById(R.id.adp_pick_round)
        pickInput = findViewById(R.id.adp_pick_in_round)
        val submit = findViewById<Button>(R.id.adp_submit_button)
        val act: Activity = this
        submit.setOnClickListener { v: View? ->
            if (playerToSearch == null) {
                generateTextOnlyFlashbar(act, "No can do", "Invalid player, use the dropdown to pick", Flashbar.Gravity.TOP)
                        .show()
                return@setOnClickListener
            }
            val roundStr = roundInput!!.text.toString()
            val pickStr = pickInput!!.text.toString()
            if (!isInteger(roundStr) || !isInteger(pickStr)) {
                generateTextOnlyFlashbar(act, "No can do", "Pick/round must be provided as numbers", Flashbar.Gravity.TOP)
                        .show()
                return@setOnClickListener
            }
            val round = roundStr.toInt()
            val pick = pickStr.toInt()
            if (pick > rankings!!.leagueSettings.teamCount) {
                generateTextOnlyFlashbar(act, "No can do", "Pick can't be higher than current league team count", Flashbar.Gravity.TOP)
                        .show()
                return@setOnClickListener
            }
            val overallPick = (round - 1) * rankings!!.leagueSettings.teamCount + pick
            hideKeyboard(act)
            getADPOddsForInput(overallPick)
        }
    }

    private fun getADPOddsForInput(pick: Int) {
        val oddsParser = ParseADPOdds(rankings, this)
        oddsParser.execute(pick, playerToSearch)
    }

    private fun setPlayerToBeChecked(id: String) {
        playerToSearch = rankings!!.getPlayer(id)
        searchInput!!.setText(playerToSearch!!.name + ": " + playerToSearch!!.position + ", " + playerToSearch!!.teamName)
    }

    private fun displayResult(output: String) {
        result!!.text = output
        val localCopy: Activity = this
        val playerId = playerToSearch!!.uniqueId
        result!!.setOnLongClickListener {
            val intent = Intent(localCopy, PlayerInfo::class.java)
            intent.putExtra(Constants.PLAYER_ID, playerId)
            startActivity(intent)
            true
        }
        clearInputs()
    }

    private fun clearInputs() {
        playerToSearch = null
        searchInput!!.setText("")
        pickInput!!.setText("")
        roundInput!!.setText("")
    }

    private fun getPlayerADPOdds(url: String, player: Player, pick: Int): String {
        try {
            val td = parseURLWithUA(url,
                    "table.table td")
            var i = 0
            while (i < td.size) {
                val possibleName = normalizeNames(td[i])
                val possiblePos = td[i + 1]
                val possibleTeam = normalizeTeams(td[i + 2])
                if (possiblePos == player.position && possibleTeam == player.teamName &&
                        (possibleName == player.name.replace("\\.".toRegex(), "") || possibleName == player.name)) {
                    return ("Odds " + player.name + " is available at pick " + pick
                            + ": " + td[i + 4])
                }
                i += 6
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get player adp likelihood", e)
        }
        return "An error occurred. Either the data is unavailable, or the internet may have dropped."
    }

    companion object {
        private const val TAG = "ADPSimulator"

        private class ParseADPOdds(private val rankings: Rankings?,
                                   private val adpSimulator: ADPSimulator) : AsyncTask<Any?, Void?, String>() {
            private val pdia: AlertDialog = MaterialAlertDialogBuilder(adpSimulator)
                    .setCancelable(false)
                    .setTitle("Please wait")
                    .setMessage("Doing fancy math...")
                    .create()

            override fun onPreExecute() {
                super.onPreExecute()
                pdia.show()
            }
    
            override fun onPostExecute(result: String) {
                super.onPostExecute(result)
                pdia.dismiss()
                adpSimulator.displayResult(result)
            }

            override fun doInBackground(vararg data: Any?): String {
                val pick = data[0] as Int
                val player = data[1] as Player
                var type = "standard"
                val scoring = rankings!!.leagueSettings.scoringSettings
                val roster = rankings.leagueSettings.rosterSettings
                if (roster.qbCount > 1 || roster.flex != null && roster.qbCount > 0 && roster.flex!!.qbrbwrteCount > 0) {
                    type = "2qb"
                } else if (scoring.receptions > 0.0) {
                    type = "ppr"
                }
                var teams = "8"
                val numTeams = rankings.leagueSettings.teamCount
                if (numTeams >= 14 && "2qb" != type) {
                    teams = "14"
                } else if (numTeams >= 12) {
                    teams = "12"
                } else if (numTeams >= 10) {
                    teams = "10"
                }
                val url = ("https://fantasyfootballcalculator.com/scenario-calculator?format="
                        + type + "&num_teams=" + teams + "&draft_pick=" + pick)
                Log.d(TAG, url)
                ParsingUtils.init()
                var first = adpSimulator.getPlayerADPOdds(url, player, pick)
                if (scoring.receptions > 0.0 && first.contains("error") && url.contains("ppr")) {
                    first = adpSimulator.getPlayerADPOdds(url.replace("ppr", "standard"), player, pick)
                }
                return first
            }
        }
    }
}